contracts/arbitrage/RevoUniswapArbitrage.sol

```diff
@@ -0,0 +1,385 @@
+        uint256[] memory _liquidityTokenBalances = new uint256[](2);
+        uint256[2] memory _initialLiquidityTokenBalances = [
+            _farmBot.stakingToken0().balanceOf(address(this)),
+            _farmBot.stakingToken1().balanceOf(address(this))
+        ];
+
+        // Get start token from sender, swap both halves for stakingToken0 and stakingToken1
+        IERC20(_args.zapToken).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _args.amountZapToken
+        );
+        uint256 _halfStartTokens = _args.amountZapToken / 2;
+
+        for (uint256 i = 0; i < _args.paths.length; i++) {
+            IERC20 _liquidityToken = IERC20(
+                _args.paths[i][_args.paths[i].length - 1]
+            );
+
+            _liquidityTokenBalances[i] = UniswapRouter.swap(
+                _farmBot.swapRouter(),
+                _args.paths[i],
+                _halfStartTokens,
+                IERC20(_args.zapToken),
+                _args.amountsOutMin[i],
+                _args.deadline
+            );
+
+            // Pre-approve the broker to spend the liquidity constituent token we swapped for
+            _liquidityToken.safeIncreaseAllowance(
+                _args.brokerAddress,
+                _liquidityTokenBalances[i]
+            );
+        }
+
+        LiquidityAmounts memory _liquidityAmounts = LiquidityAmounts(
+            _liquidityTokenBalances[0],
+            _liquidityTokenBalances[1],
+            _args.minAmountStakingToken0,
+            _args.minAmountStakingToken1
+        );
+
+        uint256 _initialFpBalance = IERC20(_args.farmBotAddress).balanceOf(
+            address(this)
+        );
+        // Mint FP using the broker; the broker will send this contract some FP and any remaining stakingToken0 and stakingToken1.
+        // any stakingToken0 and stakingToken1 returned by the broker should be sent back to the sender
+        RevoFPBroker(_args.brokerAddress).getUniswapLPAndDeposit(
+            _args.farmBotAddress,
+            _liquidityAmounts,
+            _args.deadline
+        );
+
+        // Swap the RFP for the original starting token. The caller specifies the minimum deisred amount in _args.minZapTokenOut.
+        // Since the swap will fail if the output amount is below this value, the caller is guaranteed a profit if _args.minZapTokenOut
+        // is greater than _args.amountZapToken and the function returns.
+        uint256 _amountZapTokenOut = UniswapRouter.swap(
+            _farmBot.swapRouter(),
+            _args.zapPath,
+            IERC20(_args.farmBotAddress).balanceOf(address(this)) -
+                _initialFpBalance,
+            IERC20(_args.farmBotAddress),
+            _args.minZapTokenOut,
+            _args.deadline
+        );
+
+        // Extra safety check; should not be strictly necessary if the swap router is honest
+        require(
+            _amountZapTokenOut >= _args.minZapTokenOut,
```

Comment:
not sure we need the argument _args.minZapTokenOut (since you could just compare _amountZapTokenOut to _args.amountZapToken, the amount in)

Reply:
I guess you could try to set _args.minZapTokenOut to _args.amountZapToken plus whatever gas you expect to spend on the remaining lines, though that seems hard to calculate.. unless the remaining lines were in their own method? gets a little complicated